### PR TITLE
Typed CDP Bindings - Part 2

### PR DESCRIPTION
**Pull request by bwalderman**
This changes implements a strongly-typed CdpClient object that implements the interface provided in devtools-protocol module. CdpClient callers have been updated to use the new style.

Before:
```js
await cdpServer.sendMessage({
  method: 'Target.setAutoAttach',
  params: {
    autoAttach: true,
    waitForDebuggerOnStart: false,
    flatten: true,
  },
});
```

After:
```js
await cdpClient.Target.setAutoAttach({
  autoAttach: true,
  waitForDebuggerOnStart: false,
  flatten: true,
});
```

Events:
```js
this._cdpClient.Target.on('attachedToTarget', (params) => {
  this._contextProcessor.handleAttachedToTargetEvent(params);
});
```

Command and event parameters are fully typed. TypeScript interfaces for CDP types are also available from the `Protocol` namespace. e.g.: `Protocol.Target.TargetInfo`. Some functions that accept protocol types like TargetInfo as a parameter have been updated to use these.

The devtools-protocol module only provides types, and no implementation, so the implementation is up to us. The new code in cdpClient.ts reads the protocol .json files and uses this information at runtime to generate classes that implement the expected interface. When a generated method on a generate class is called, it will send a raw CDP message using a CdpTransport object and await a response.

Follow-up:
- There are still some manually-written CDP calls in mapperServer.ts that can be updated to use the new CdpClient. Needs a CdpTransport for websocket.
- Unit test coverage for CdpClient. Waiting for #12

### CODE REVIEW

File: src/bidiMapper/commandProcessor.ts
Comment (sadym-chromium) [inline src/bidiMapper/commandProcessor.ts:59]: This method was intended to inject all the message listeners logic and later to be split between different domains.

File: src/bidiMapper/mapper.ts
Comment (sadym-chromium) [inline src/bidiMapper/mapper.ts:71]: `WindowGlobalCdpTransport` -> `WindowCdpTransport `
```suggestion
  class WindowCdpTransport implements CdpTransport {
```

File: src/bidiMapper/mapper.ts
Comment (sadym-chromium) [inline src/bidiMapper/mapper.ts:87]:
```suggestion
  return connectCdp(new WindowCdpTransport());
```

File: src/bidiMapper/tsconfig.json
Comment (sadym-chromium) [inline src/bidiMapper/tsconfig.json:6]: +1

File: src/bidiMapper/utils/cdpClient.ts
Comment (sadym-chromium) [inline src/bidiMapper/utils/cdpClient.ts:25]: I like the approach with dynamic generating methods, but it can confuse IDE. Alternative I can think of is generating a TS file based on JSON. Not sure if we need it now though. WDYT?

File: src/browserLauncher.ts
Comment (sadym-chromium) [inline src/browserLauncher.ts:110]: +1

**Review by sadym-chromium — APPROVED** [review]

**Comment by sadym-chromium** [comment]
I'll wait with merging #20, as it can involve lots of conflicts. So please go ahead and merge this one, and I rebase and merge #20 afterwards (after getting LGTM, of cause).

File: src/bidiMapper/utils/cdpClient.ts
Comment (bwalderman) [reply src/bidiMapper/utils/cdpClient.ts:25]: Right now, the IDE uses the pre-existing types from https://github.com/ChromeDevTools/devtools-protocol/tree/master/types. The nice thing about using pre-built type definitions is that all of these types are known at compile time. The drawback is there is no implementation and we need to generate the implementation at runtime. If there's a bug in how we generate the implementation, then the IDE's understanding of the types won't match reality. I think this code is correct though :)

If we generate our own typescript files that include an implementation, it will ensure the "compile-time" and "runtime" types always match, but we'll need an extra pre-build step to generate the files. The IDE won't recognize any CDP types until the developer builds at least once.

It might be useful to do this in the future if we need more control or want to have a different API than the one defined in `devtools-protocol`. For now though, I think dynamically generating the methods is good enough.

File: src/bidiMapper/commandProcessor.ts
Comment (bwalderman) [reply src/bidiMapper/commandProcessor.ts:59]: I think we can still organize this so that different domains are separate. I'm imagining `BrowsingContextProcessor` can have a public method like `registerEventListeners(cdpClient)` that we call from here. The `registerEventListeners` method can register for whatever events the `BrowsingContextProcessor` needs, and the `CommandProcessor` doesn't need to know or care.

**Comment by sadym-chromium** [comment]
I'm merging the change to unblock myself with the #20

**Pull request merged**
